Issue790: MultiGet-MRR implementation

spetrunia opened the pull request:
First implementation.
Supports scans on full primary key
Cost-based choice whether to use MRR/BKA is not supported (set mrr=on,mrr_cost_based=off)

spetrunia committed:
Issue #790: MultiGet-based MRR implementation

- Initial implementation with a basic testcase
- Uses STL containers for buffers, ignoring the provided buffer space
  and doesn't have any limit on batch size (well, SQL layer would limit
  it due to its own use of buffer)
- Cost-based choice whether to use MRR is not present.

spetrunia committed:
Issue #790: MultiGet-based MRR implementation, part 2

- More test coverage
- Support HA_MRR_SUPPORT_SORTED

spetrunia committed:
Code cleanup

spetrunia committed:
Issue #790: MultiGet-based MRR implementation, part 4

- Switch to using the variant of MultiGet() that uses C-style arrays
- Use the buffer provided as an argument to multi_range_read_init()
 - The buffer is used to store
   = keys
   = rocksdb::Slice describing the keys
   = Return value objects
   = return value statuses
   = the values of KEY_MULTI_RANGE.ptr
 - The memory to store the returned records is still allocated inside
   RocksDB somehow (not sure if this is even avoidable?)

spetrunia committed:
Issue #790: MultiGet-based MRR implementation, part 5

- It turns out multi_range_read_info_const() should provide buffer
  size estimates, after all.
- Handling for edge cases (when ranges are exhausted right when the
  buffer is filled)

spetrunia committed:
Issue #790: MultiGet-based MRR implementation, part 6

- For very small buffer sizes, fallback to default MRR implementation.
- Handle statements that do writes.

spetrunia committed:
Fix a typo in previous commit

spetrunia committed:
More test coverage

spetrunia committed:
Issue #790: MultiGet-based MRR implementation: on/off switch

MultiGet MRR or BKA should work only when

  mrr_cost_based=off,nrr=on

The default is

   mrr=on,mrr_cost_based=on

so it is not enabled by default

spetrunia committed:
Issue #790: Coverage for status counters

spetrunia committed:
Apply the provided formatting diff

spetrunia committed:
Issue #790: MultiGet-based MRR: Support secondary indexes

Support secondary indexes. For non-index_only scans, we can
use the MultiGet to read the clustered PK records.

Under the hood, the difference between scan on clustered PK
and secondary index is as follows:
- for Clustered PK, we get the PK values from SQL layer
- for other indexes, we get the PK values from doing an index-only
  scan on the secondary index.

spetrunia committed:
Issue #790, MyRocks/MRR: make ref(const) access switch to MRR

The optimizer didn't use MRR for ref(const) access. It used
h->index_read_map(); h->index_next_same() calls instead.

This patch adds a heuristic: if we picked ref(const) and
the best quick select used the same index and it used engine's
MRR implementation - switch to using the quick select which would
use MRR.

Weak points:
- if we picked ref(const) on index I1 while the best quick select
was using another index I2, we would not switch to using MRR.
(note that inconsistencies between quick select cost and ref access
costs are a known weak point).

spetrunia committed:
Issue #790: MultiGet-based MRR, Secondary Keys

When scanning a secondary index, mrr_n_ranges is not the same
as #rowids.
Introduce mrr_n_rowids. Note that we don't have a reliable
estimate of how many rowids we will have.
(TODO: can we get rid of array-based allocations in form
   mrr_statuses = new (buf) rocksdb::Status[n_elements];
and invoke the constructors when we actually need them?)

spetrunia committed:
Issue #790: MultiGet-based MRR, Secondary Keys

- Fix an issue where one would switch off MRR and then get wrong query
  results: if MRR code has set m_keyread_only, it should also clear it.

- Make Index Condition Pushdown work.

spetrunia committed:
Post-rebase fixes

spetrunia committed:
Issue #790: MultiGet-based MRR: Check if the query has been killed

The loop in mrr_fill_buffer() should have a check whether the query
has been KILL-ed.

yoshinorim commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15118,373 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    range_seq_t seq_it;
+    seq_it = seq->init(seq_init_param, n_ranges, *flags);
+    while (!seq->next(seq_it, &range)) {
+      if (!(range.range_flag & UNIQUE_RANGE)) {
```
Could you check if thd is marked as killed or not in the loop, and returning errors if killed, like this?

```
      if (thd && thd->killed) {
        return HA_ERR_QUERY_INTERRUPTED;
      }
```

yoshinorim commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15118,373 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+  int key_size;
+  char *range_ptr;
+  while ((key_size = mrr_rowid_reader->get_next_rowid((uchar*)buf, &range_ptr)) > 0 ) {
+
```
thd->killed check, same as above

spetrunia replied:
Done

yoshinorim commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15118,373 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+  }
+
+  while (1) {
+    if (mrr_read_index >= mrr_n_elements) {
```
thd->killed check, same as above

spetrunia committed:
Issue #790: MultiGet-based MRR: Code cleanup

Code cleanup.

spetrunia committed:
Issue #790, MyRocks/MRR: make ref(const) access switch to MRR

Followup patch: update test results:

- index_merge_innodb uses a table that's large enough so that the
  choice to use MRR is cost-based
- Other tests set mrr=on,mrr_cost_based=off, which forces MRR to be used.

spetrunia committed:
Issue #790, MyRocks/MRR: Fix a problem with ORDER BY queries

SQL layer can request Sorted Mode, even if we indicated we don't support
it. When this happens, the expected action is to switch to the default
MRR implementation.

(EXPLAIN will still work correctly, see the issue comment on github for
explanations how)

spetrunia committed:
Issue #790, MyRocks/MRR: make ref(const) access switch to MRR

Revert this (as a different solution will be used):

Followup patch: update test results:
- index_merge_innodb uses a table that's large enough so that the
  choice to use MRR is cost-based

- Other tests set mrr=on,mrr_cost_based=off, which forces MRR to be used.

spetrunia committed:
Issue #790, MyRocks/MRR: make ref(const) access switch to MRR

Do the switch only if the storage engine requests it by setting
the HA_MRR_CONVERT_REF_TO_RANGE flag.

spetrunia committed:
Fix issues found by RQG

- As the comment in multi_range_read_init() already says, it should
  not return HA_ERR_END_OF_FILE. This confuses the SQL layer

- The rows are returned in mrr_values[] array, and we need the current
  one in m_retrieved_record. Do "assignment" of PinnableSlice objects
  the right way.

spetrunia committed:
Issue #790, MyRocks/MRR: fix issues found by RQG

In ha_rocksdb::read_row_from_secondary_key(), get the key/value from
the iterator AFTER find_icp_matching_index_rec() has navigated to the
index tuple that matches the ICP.

spetrunia committed:
Issue #790, MyRocks/MRR: code cleanup

- Mrr_sec_key_rowid_source::get_next_rowid() does an index scan
on the secondary index. This may produce an error (e.g. "data
corrupted" or other kind of RocksDB). Don't swallow this error,
return it all they way up to the SQL layer.
- Do TTL filtering (if we find an expired row, skip it).

spetrunia commented:
Ok,  now the code is ready for the review.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    }
+
+    if (all_eq_ranges) {
+      // Indicate that we will use Mutlit-Get MRR
```
"Multi-Get" or "MultiGet"?

spetrunia replied:
Fixed

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    if (all_eq_ranges) {
+      // Indicate that we will use Mutlit-Get MRR
+      *flags &= ~HA_MRR_USE_DEFAULT_IMPL;
+      *flags |= HA_MRR_CONVERT_REF_TO_RANGE;
```
Does the `HA_MRR_CONVERT_REF_TO_RANGE` flag make sense for primary keys? It's not like we're going to use MRR with ref plans on primary key anyway?

spetrunia replied:
You're right, this flag won't make any difference here. Lookups on full constant PK will use eq_ref, not ref...

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+        }
+      }
+      // Skip the "is not found" errors
+      if (mrr_statuses[mrr_read_index].ok()) break;
```
Perhaps we should check explicitly for "not found" vs other errors here.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    }
+
+    rc = convert_record_from_storage_format(&rowkey, table->record[0]);
+    m_retrieved_record.Reset();
```
Is this safe to do for TEXT/BLOBs? From what I remember, we cannot call Reset right away because it may be pointing to blob data.

If you look at the existing usage of PinnableSlice, we just call Reset right before we issue a Get:
https://github.com/facebook/mysql-5.6/blob/7bdd211c8eb75740dc881289ef3b3514235f432c/storage/rocksdb/ha_rocksdb.cc#L3466

spetrunia replied:
Agree. Changed accordingly.

lth replied:
Maybe I missed it ... but I don't actually see where this was addressed? (or maybe it wasn't an issue after all?)

spetrunia replied:
Yes... I remember addressing it, but the change is not there... will fix now

spetrunia replied:
Ok, I've added the change. I wasn't able to construct a testcase for this, though - I couldn't find a way to cause RocksDB make the underlying data be released before the SQL layer tries to access it.  The code looks fairly clear, though.

spetrunia replied:
(now the testcase is constructed and pushed)

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3781,6 +3791,24 @@ class Rdb_writebatch_impl : public Rdb_transaction {
+                 const size_t num_keys, const rocksdb::Slice *keys,
+                 rocksdb::PinnableSlice *values, rocksdb::Status *statuses,
+                 bool sorted_input) override {
+    // todo: could we just read the committed content from the DB here?
```
What's wrong with just calling `MultiGetFromBatchAndDB` like below? (That said, I don't think this code path will be exercised).

spetrunia replied:
Nothing is wrong, and I think we can do with just one `multi_get` function now (since the bypass' function switched to using the same function signature.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -7991,8 +8022,6 @@ int ha_rocksdb::read_row_from_secondary_key(uchar *const buf,
 
-  /* Get the key columns and primary key value */
-  const rocksdb::Slice &rkey = m_scan_it->key();
   const rocksdb::Slice &value = m_scan_it->value();
```
`covered_lookup` uses  `kd.covers_lookup` which determines results on a per-value basis. Currently, it looks like we only calculate `covered_lookup` once.

That said, the covered key feature (https://github.com/facebook/mysql-5.6/pull/639) is currently off, but it's probably easier/safer just to fix this.

spetrunia replied:
Yes, fixed.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    key_part_map all_parts_map =
+        (key_part_map(1) << self->m_pk_descr->get_key_parts()) - 1;
+    DBUG_ASSERT(range.start_key.keypart_map == all_parts_map);
+    DBUG_ASSERT(range.end_key.keypart_map == all_parts_map);
```
Might as well assert on flags as well?

spetrunia replied:
Added asserts.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+  (*) The memory for rows is allocated somewhere inside RocksDB, there's no
+      way to make it use the user-supplied buffer.
+  (**) The engine could specify HA_MRR_NO_ASSOCIATION which would mean "we
+      cannot tell which key the returned records match" we don't do this.
```
It seems like we do have some sort of handling below for `HA_MRR_NO_ASSOCIATION`, and this flag gets passed in `get_quick_select_for_ref` which seems to be used only for filesorts? (I couldn't tell which plans called `get_quick_select_for_ref` though)

spetrunia replied:
`get_quick_select_for_ref` is called when one has
- ref(const) access on the first table 
- "Using filesort" (w/o "Using temporary).

In this case, the query plan is to get filesort() to read the first table and
sort it according to the sort criteria.  filesort is unable to use ref access
as the input stream, it can either read from a quick select or do a full table
scan. Because of that, get_quick_select_for_ref is used.

spetrunia replied:
The code here doesn't really handle `HA_MRR_NO_ASSOCIATION`. We just increment the counter, in the same way like DsMrr_impl::dsmrr_init  does:
```
  is_mrr_assoc= !MY_TEST(mode & HA_MRR_NO_ASSOCIATION);
  if (is_mrr_assoc)
    status_var_increment(table->in_use->status_var.ha_multi_range_read_init_count);
```
This allows to use the counter value as a counter of how many times BKA was used.

Some background:
Initially, HA_MRR_NO_ASSOCIATION was introduced for MRR/NDB . NDB returns records but it cannot tell which lookup key each record is for. 

Now, HA_MRR_NO_ASSOCATION is used as follows:
* BKA code does need the storage engine to tell which lookup key the record is for. It doesn't pass HA_MRR_NO_ASSOCATION. 
* Range access doesn't need this, it just needs the record. It passes HA_MRR_NO_ASSOCATION flag.

So, for an MRR implementation, HA_MRR_NO_ASSOCATION means that it doesn't need to keep track which result record matches which range. For  MyRocks/MRR, we won't need the `mrr_range_ptrs` array.  This allows to save a bit of buffer space (8 bytes per range) ...  but you can only do that when you are using range access, and typically don't have many ranges.

Should I implement HA_MRR_NO_ASSOCATION still?

hermanlee replied:
It doesn't seem like supporting HA_MRR_NO_ASSOCIATION is really needed? It seems to be passed in mainly as an optimization to save on the space? It looks like the bka_unique_range_seq_skip_record callback requires !NO_ASSOCIATION and that appears to be used mainly for filtering. 

I think we don't need to implement the HA_MRR_NO_ASSOCIATION feature buffer unless saving 8 bytes per range is really critical.

lth replied:
> This allows to use the counter value as a counter of how many times BKA was used.

Ah okay, this makes sense a lot more sense now. We're just using `HA_MRR_NO_ASSOCATION` to tell us if BKA was used or not (and the Handler_mrr_init counter is meant to only count MRR used for BKA).

I don't think we need to implement `HA_MRR_NO_ASSOCIATION` either. I was just curious why we bumped this counter only sometimes with `HA_MRR_NO_ASSOCIATION`.

spetrunia replied:
Ok

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+  Mrr_sec_key_rowid_source(ha_rocksdb *self_arg) : self(self_arg), err(0) {
+  }
+
+  int init(RANGE_SEQ_IF *seq, void *seq_init_param,
```
Do we have to handle the callbacks passed into `RANGE_SEQ_IF`, like `skip_index_tuple` or `skip_record`.

spetrunia replied:
Yes for `skip_index_tuple`, no for `skip_record` but I've added handling for both  ( notes taken during investigation are here: https://github.com/facebook/mysql-5.6/issues/790#issuecomment-541188976)

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+    range_seq_t seq_it;
+    seq_it = seq->init(seq_init_param, n_ranges, *flags);
+    while (!seq->next(seq_it, &range)) {
+      if (!(range.range_flag & UNIQUE_RANGE)) {
```
Maybe add an assert that it is EQ_RANGE too?

spetrunia replied:
In my opinion, it would be redundant. Looking at the definition of EQ_RANGE and UNIQUE_RANGE in include/my_base.h, one can see that UNIQUE_RANGE is a stronger requirement than EQ_RANGE.
The code that sets these flags in opt_range.cc is fairly straightforward so I don't think there's any risk of getting a UNIQUE_RANGE that's not a EQ_RANGE.

lth reviewed (commented):
It's kind of strange how the same sql level switch will give different behaviour on innodb vs rocksdb tables. I wonder if we want to add an extra rocksdb switch for people running with two engines. At least for us, we mainly run with single engine so it's not a big deal.

spetrunia committed:
Issue #790, MyRocks/MRR: address review input and cleanup

- Join two multi_get() functions together (the other implementation
  that is used by bypass code now uses RocksDB's MultiGet with the same
  signature, so there's no difference)
- Add more DBUG_ASSERTs as requested
- Call RANGE_SEQ_IF::skip_index_tuple when appropriate (+testcase)
- Call RANGE_SEQ_IF::skip_record when appropriate (+testcase)

spetrunia committed:
Issue #790, MyRocks/MRR: address more review input

Fix the code ha_rocksdb::read_row_from_secondary_key(). The idea is
that index-only scans can now Pushed Index Conditions (ICP):
one can run MRR scan with ICP, and MyRocks/MRR will use an index-only
scan under the hood to collect the rowids.

Previous variant of the code broke the "covered lookups" feature.

spetrunia commented:
@lth , I have addressed (or replied to) all the input.  The patch is ready for another review pass.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15089,497 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+          return HA_ERR_END_OF_FILE;
+        }
+
+        if (table->in_use->killed) return HA_ERR_QUERY_INTERRUPTED;
```
This should be checked on each row being processed, but it's a minor point.

spetrunia replied:
As far as I understand, there is a check for `thd->killed` for each row processed (actually, there are several). 
My reasons for putting the check into this loop were:
* the loop iterates through the arrays returned by MultiGet(), so it enumerates O(#rows_read). 
* Some subtle bug could cause execution to get stuck in the loop (after all, there are several "continue;" statements in this loop).

Should this check be kept or removed?

hermanlee replied:
We should keep the check, but I was thinking it should be moved outside of the 'if (mrr_read_index >= mrr_n_elements) conditional. If, for whatever reason, mrr_read_index is never incremented or mrr_n_elements is really large, we can still loop without hitting this check.

Maybe move the check right after the second while (1)?

Also, is it possible when executing query plans for table->in_use to not be the same as ha_rocksdb::ha_thd()?

spetrunia replied:
> is it possible when executing query plans for table->in_use to not be the same as ha_rocksdb::ha_thd()?

As far as I know, no.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -10560,6 +10569,8 @@ int ha_rocksdb::index_end() {
   active_index = MAX_KEY;
   in_range_check_pushed_down = FALSE;
 
+  mrr_free();
```
Minor point: can this be conditioned on mrr_uses_default_impl? Otherwise, it seems like every index_end() call always goes through mrr_free(), even though mrr is never invoked. It might be safer to always call mrr_free(), but it seems unnecessary most of the time.

spetrunia replied:
Yes. I think it's better to condition on `mrr_rowid_reader`  (which is always present when using a non-default MRR implementation). Done.

spetrunia committed:
Issue #790, MyRocks/MRR: address review input part 3

- in ha_rocksdb::index_end(), call mrr_free() only if we are in an MRR scan.
- added a comment.
- Move the "if (table->in_use->killed) return HA_ERR_QUERY_INTERRUPTED" logic

spetrunia committed:
Issue #790, MyRocks/MRR: address more review input: don't release m_retrieved_record

Keep the m_retrieved_record pinned.

spetrunia committed:
Issue #790, MyRocks/MRR: add test coverage for previous cset

lth commented on mysql-test/suite/rocksdb/t/rocksdb_mrr.test:
```diff
@@ -0,0 +1,309 @@
+
+--disable_warnings
+drop table if exists t0,t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11;
+--enable_warnings
```
We don't use the `drop table if exists` pattern anymore because mtr should ensure correct cleanup for us (and also for other tests).

spetrunia replied:
Removed.

lth reviewed (commented):
One minor issue: It looks like we're not updating the rows read counters (both the status var and the counters for user/table stats). Can we update the counter as well?

It's also unfortunate that we'd have to pin our data in block cache until the mrr buffer finishes reading just because of blobs, but this doesn't seem to be an issue for now.

lth commented:
Some other stats may be missing as well (eg. rows_requested)

spetrunia committed:
Issue #790, MyRocks/MRR: properly increment per-table read counters

- Properly increment I_S.table_statistics.rows_* counters
- Do release the PinnableSlice that's holding the retrieved record
  as soon as we don't need it (before, we released all records in
  ha_rocksdb::mrr_free_rows(). Now, we only release them there if
  the mrr scan was not finished).
- Clean up in testcases

lth commented on mysql-test/suite/rocksdb/t/rocksdb_mrr.test:
```diff
@@ -0,0 +1,389 @@
+--echo #
+--echo # Check the userstat counters
+--echo #
+show variables like '%user_table_stat%';
```
This isn't a useful variable to check. This controls the population of the `information_schema.user_table_statistics` table (not table_statistics). We can probably just remove this.

spetrunia replied:
Removed

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -15076,6 +15094,513 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+
+    m_retrieved_record.Reset();
+    m_retrieved_record.PinSlice(mrr_values[cur_key], &mrr_values[cur_key]);
+    mrr_values[cur_key].Reset();
```
It looks like this was newly added, but I don't think it actually does anything.

In 
```
    m_retrieved_record.PinSlice(mrr_values[cur_key], &mrr_values[cur_key]);
```

We're calling

```
  inline void PinSlice(const Slice& s, Cleanable* cleanable) {
    assert(!pinned_);
    pinned_ = true;
    data_ = s.data();
    size_ = s.size();
    cleanable->DelegateCleanupsTo(this);
    assert(pinned_);
  }
```

and `DelegateCleanupsTo` will actually set the cleanup functions to nullptr on the origin slice. Effectively, we've already "move"d the Slice away, so `mrr_values[cur_key] ` should be empty already.

spetrunia replied:
Agree. I misread what PinnableSlice is doing. Reverted the change

lth reviewed (commented):
I think we're still missing the call to `update_row_stats` for ROWS_READ. This will update the `rocksdb_rows_read` status var.

spetrunia committed:
Cleanup tests

spetrunia committed:
Issue #790, MyRocks/MRR:

- Increment the rocksdb_rows_read counter. It shows the number of rows
  that the SQL layer has read.

- Remove attempted "optimization" with mrr_values[cur_key].Reset() call.
  It didn't help anything, actually.

spetrunia commented:
Hi @lth,

Thanks for the input so far,

I've followed it and made `rocksdb_rows_read` to show the number of rows the SQL layer has read from MyROcks.

However, I've previously made `I_S.table_statistics.rows_read` to be the number of rows that MyRocks has read from RocksDB:
 https://github.com/spetrunia/mysql-5.6/blob/fb-mysql-5.6.35-issue790/storage/rocksdb/ha_rocksdb.cc#L15518

These two numbers are typically the same, but they can differ if the SQL layer stops the MRR scan before it has read all the rows.

An example of this can be found here:
https://github.com/spetrunia/mysql-5.6/blob/fb-mysql-5.6.35-issue790/mysql-test/suite/rocksdb/r/rocksdb_mrr.result#L385
It shows that for table t2, I_S.table_statistics.rows_read=10 while t2's contribution to ROCKSDB_ROWS_READ is 2.

Is the current state fine, or the counters should be made to agree? If we need them to agree, which one should be changed?

lth commented:
@spetrunia Thanks, the current implementation is fine.
